### ISSUE TITLE
[Feature Request] Scrolling up and down using the mouse scroll wheel

### ISSUE THREAD

**Issue by sianhulo** [nomacs/nomacs#237]
Instead of using it to zoom in and out, I'd prefer to be able to map that to ctrl+scrollwheel and use scrollwheel to pan the image up and down.

**Comment by diemmarkus** [comment]
sorry, but the wheel functions are pretty crowded already (CTRL+wheel switches between images) - so I guess we won't add this feature...

**Comment by aspiers** [comment]
@diemmarkus please could you reconsider your position based on the following facts?

1. Almost all users will expect the scroll wheel to scroll, not zoom.  This is the universal convention.  That's why it's called a scroll wheel not a zoom wheel ;-)

2. Control+wheel is an extremely common convention for zooming.

3. Shift+wheel is an extremely common convention for horizontal scrolling.

4. The argument "wheel functions are pretty crowded already" doesn't really apply, because there are plenty of modifiers to allow the mouse wheel to perform multiple functions.  In addition to Control and Shift, there's Alt, Control+Shift, Control+Alt, Shift+Alt, and even Control+Shift+Alt.

No single mapping will make all users happy, therefore the most logical thing to do is to:

1. conform with the universal conventions on mouse wheel bindings listed above, and

2. make them all configurable so that users can always set it up the way they prefer.

**Comment by sianhulo** [comment]
that's right, personally I'm happy even if these settings aren't made default, just being able to bind it would be more than enough

**Comment by diemmarkus** [comment]
Hi @aspiers,

I think what users expect is heavily depending on the OS they are using (nomacs tries to bridge that). When thinking about user experience _extremely common_ is not what you want in general. For example placing the back button at the top left corner is extremely common - still it's not a good practice since this corner is hard to reach.

> Almost all users will expect the scroll wheel to scroll, not zoom. This is the universal convention. That's why it's called a scroll wheel not a zoom wheel ;-)

let's call it *mouse wheel* : ) - Take a look at [Google Maps](https://www.google.com/maps/) - it has a pretty similar zoom/pan behavior as nomacs and is used by _a lot_ of users.

> conform with the universal conventions on mouse wheel bindings listed above, and

the thing here is, that it makes sense for text editors etc. to scroll using the mouse wheel since their content is mostly vertical. However for image viewers this is not the case (the mouse wheel would do nothing if the image fits the window). Most image viewers I know (including IrfanView, Windows Photo Viewer, ...) use the mouse wheel for switching between images. This behavior can be configured in nomacs too.

best,
markus

**Comment by sianhulo** [comment]
I'm sorry, but if I found this could be configured from nomacs I wouldn't have opened this issue. Where's the option? I found the option to have the scroll wheel NOT zoom, and despite zoom in and zoom out being in keyboard shorcut, I can't map it to ctrl+scroll as it simply doesn't seem to "grab" the scrolling event.

**Comment by diemmarkus** [comment]
sorry, that was misleading. When I wrote

> This behavior can be configured in nomacs too.

I meant the option to switch between images with the mouse wheel in stead of zooming. The behavior you requested (panning using the wheel) is not configurable in nomacs (yet : ).

**Comment by aspiers** [comment]
Thanks both for the good discussion.

[**@diemmarkus**](https://github.com/diemmarkus) commented on [8 Aug 2018, 15:33 BST](https://github.com/nomacs/nomacs/issues/237#issuecomment-411427661 "2018-08-08T14:33:44Z - Replied by Github Reply Comments"):
> The behavior you requested (panning using the wheel) is not configurable in nomacs (yet : ).

This "yet" sounds more encouraging - I like it ;-)

Our discussion re-affirms my original assertion:

> No single mapping will make all users happy

so I'd like to re-iterate:

> therefore the most logical thing to do is to:
>
> [...]
>
> - make them all configurable so that users can always set it up the way they prefer.

In almost every other way nomacs is awesome, but the lack of convenient panning (see also #241 and #242) is a deal-breaker feature for me.

**Comment by diemmarkus** [comment]
I'll keep you posted when we have progress here...

> In almost every other way nomacs is awesome,

thanks : )

**Comment by aspiers** [comment]
[**@diemmarkus**](https://github.com/diemmarkus) commented on [9 Aug 2018, 12:34 BST](https://github.com/nomacs/nomacs/issues/237#issuecomment-411727593 "2018-08-09T11:34:51Z - Replied by Github Reply Comments"):
> I'll keep you posted when we have progress here...

I appreciate it, thanks!

**Comment by mdmawat** [comment]
Disabling "Mouse Wheel Zooms" doesn't have Ctrl+wheel work on zooming and I finally found Alt+wheel works. Any way to have zooming functionality switched to Ctrl+wheel?

**Comment by honza-zidek** [comment]
For me this issue is clearly a showstopper. I switched to nomacs from Irfanview (licensing issue, not functional) and am considering to stop using nomacs at all. This ignoring all the conventions and not allowing the users to configure the mouse behaviour themselves makes nomacs almost unusable. Also the lack of horizontal and vertical scrollbar makes the life much harder.
You won't make a software great by making it counter-intuitive and standing against the common ways.

**Comment by diemmarkus** [comment]
> Also the lack of horizontal and vertical scrollbar makes the life much harder.
you can enable scrollbars in `Settings > General`.

> You won't make a software great by making it counter-intuitive and standing against the common ways.
thanks - now I know...

**Comment by honza-zidek** [comment]
> > Also the lack of horizontal and vertical scrollbar makes the life much harder.
> 
> you can enable scrollbars in `Settings > General`.
> 

Well, finally I found it in `Settings > Display`, not in `Settings > General`, but thanks for the hint that such an option exists. I would not expect anything like this to be by default `off`.

> > You won't make a software great by making it counter-intuitive and standing against the common ways.
> 
> thanks - now I know...

Well, based on the feedback you have from the **users of *your* program** you might be less sarcastic and more thoughtful. Anyway, it's your program, your users and your decision, so you may do with us whatever you please. We use it for free, so we should just shut up and be grateful to The Creator.

**Comment by dabalciunas** [comment]
> For me this issue is clearly a showstopper. I switched to nomacs from Irfanview (licensing issue, not functional) and am considering to stop using nomacs at all. This ignoring all the conventions and not allowing the users to configure the mouse behaviour themselves makes nomacs almost unusable. Also the lack of horizontal and vertical scrollbar makes the life much harder.
> You won't make a software great by making it counter-intuitive and standing against the common ways.

I have to agree with him, you guys are really missing the point here. If the default behavior of nomacs is like you suggest us to use, no problem at all. But you should at least provide the users a way to customize nomacs for setting up the "conventional" shortcuts behavior, which is offered by so many tools in the market. I'm a software developer as you guys, and I had to learn this lesson too.

**Comment by supersexy** [comment]
Can confirm that first experience of nomacs made me conclude "not usable" because of not finding a way to pan the picture neither by keyboard or by mouse and not finding any hint about keyboard shortcuts and no way to configure these absolutely basic functions. 

It is an example for a good software that unfortunately fails in delivering totally expected very basic function and makes it very hard to try it a second time because the "very basic" category does not work as expected. 

It is like yuo artificially added an anti-feature to the software to make people avoid using it. 

Please test other image viewers and you will see that it usually is very easy to pan / navigate / scroll etc - these set of functions should be undertandable for users without even thinking about it. 

Please learn from these comments, as I still see lots of potential.

**Comment by honza-zidek** [comment]
> Can confirm that first experience of nomacs made me conclude "not usable" because of not finding a way to pan the picture neither by keyboard or by mouse and not finding any hint about keyboard shortcuts and no way to configure these absolutely basic functions.
> 
> It is an example for a good software that unfortunately fails in delivering totally expected very basic function and makes it very hard to try it a second time because the "very basic" category does not work as expected.
> 
> It nearly feels like you added an anti-feature to the software on purpose to make people avoid using it, it is something where people get crazy asking "how can a developer not understand that?".
> 
> Please test other image viewers and you will see that it usually is very easy to pan / navigate / scroll etc. - these set of functions should be undertandable for users without even thinking about it. But this happens because developers go deep into these details, so should you.
> 
> Please learn from these comments, as I still see lots of potential.

I totally agree with you. I'm afraid it's a common problem of opensource software made by programmers for themselves and not with the regard of the users. Big potential lost in user-unfriendly  GUI. On the other side of the spectrum there are big corporations, where product managers force the programmer to produce software which technically is crap, but is created for users. It's difficult to decide who is worse - the stubborn technician ("I don't care what other similar software looks like, the way of mine is the best") or the manager ignoring technical issues ("I don't care that the database model is not designed for this particular feature, I simply must have it there").

**Comment by AnonymousPomp** [comment]
One more person who agrees with @aspiers

**Comment by Tromick** [comment]
I installed Irfanview in 2009 but uninstalled it after 5 min due of it wouldn't allow to change shorcuts. I installed it after 11 years, people are begging for this option and still no user-friendly. No wonder it's always criticized by program editors. Also it doesn't show images sharp if it is windowed mode, my screenshot's texts were so blurry in windowed mode. Probably they will listen users in 2030 or more far future.

**Comment by honza-zidek** [comment]
@Tromick do you really mean Irfanview?

**Comment by diemmarkus** [comment]
I am pretty sure that nomacs did not exist in 2009 : )

**Comment by Tromick** [comment]
> I am pretty sure that nomacs did not exist in 2009 : )

Nomacs is pretty good, i really loved it, thank you. Is there way to close tab thing and open it separately like there is setting in Adobe PDF. Also i wish that doesn't playing GIF's faster. This is why i was installed Irfanview back but it was horrible experience same as in 2009.

**Comment by sianhulo** [comment]
might have something to do with the delay between frames, a lot of gifs are set to 0, which would make the gif play instantly, so it's up to the client to set its own delay. Basically if the gif has delay 0 it's the problem of the image.

**Comment by lukasbloom** [comment]
@aspiers +1

**Comment by Tojaw** [comment]
Is there any update on this? I'd like to be able to use simple scrolling to scroll, Ctrl + scroll to zoom, Shift + scroll to scroll horizontally (it isn't important to me since I have a horizontal wheel, but I'm pretty sure it's important to many) and Alt + scroll to switch between images.

My suggestion would be to add an option where you are presented with the four different scroll actions and where you can assign buttons to all of them.
E.g:

- Switch between images: None
- Zoom: Ctrl
- Scroll vertically: Alt
- Scroll horizontally: Shift

This would be the default, and I can freely change these settings, disable some actions or maybe even assign multible possible buttons to one action.

**Comment by honza-zidek** [comment]
> Is there any update on this? 

@Tojaw I found the update long time ago. Its name is FastStone Image Viewer :) I am a happy user of this and the nomacs and its stubborn developer can still happily live in his own world of weird mouse controls :)

**Comment by hced** [comment]
No need for all the immature, unthankful ways to express your feelings (some of you).

@Markus (@diemmarkus) I love your program, heaps! Would you please consider just chiming in here (disregarding all the immature commenters) with a final say in this matter. Now, I myself think from a user standpoint that a personal choice (given how split we are in this regard) would be the right move here. You're totally right, MWHEEL-scrolling is indeed used for Zooming in some apps, in many cases. As mentioned, there are also cases where it's used for Prev/Next file in other apps (eg. PDF readers). I think that discussion is pretty fruitless.

Instead, I'd like to kindly ask if this is something that's completely unthinkable? Or would you be so **awesome** that you might consider fulfilling our wishes? Whatever it's going to be, I highly value all the work you've put into making this fine app. 

Please take a bit of think. Keep us informed when you've come to a definite decision. Take care and best wishes!!

**Comment by cjrobe** [comment]
I was going to use Nomad to preview images in my workflow in between using graphics editors. The fact that Nomad won't function in any way except counter to every single popular graphics editor makes in a non-starter for me. Uninstalled in two minutes. Shame because the program appears to be quite nice other than this.

**Comment by hced** [comment]
My suggestion to developers like @dabalciunas: Since you're a programmer and show interest in Nomacs. Can't you just implement the choice and send him a Pull Request? That might be the tipping point.

**Comment by Gaurav-049** [comment]
So, we will never get option to assign two finger touch up & down gesture to scroll/pan instead of image browsing?

**Comment by jammet** [comment]
I keep coming back to this year after year, and nomacs would in theory be my perfect image viewer. But I just cannot go without using the mouse wheel for going to the previous or next image. How often in my viewing do I need zoom? Considering? Pretty often. But I flick through the images so much more often, that there is just no comparing. And my default is typically to just fit the image to the window or fullscreen dimensions, either by blowing it up, or shrinking it.

Please consider adding this as a configurable option, in the future. Much love.

**Comment by scrubbbbs** [comment]
> But I just cannot go without using the mouse wheel for going to the previous or next image

This issue is about panning/scrolling the image, not previous/next image, which is supported.

There are two defaults that must be unchecked:

- **Mouse Wheel Zooms**
- **Next Image on Horizontal Zoom**

This will swap the function of the the scroll wheel and back/forward mouse buttons.

If this does not work for you need to install the current beta (I can't say if 3.16/3.17 had this.. maybe maybe not)

**Comment by ted423** [comment]
To be honest, the instructions in the settings make it really hard to understand why it has to be done this way. I wish it were the other way around, so I wouldn't have to configure it for others after installing it for them.

**Comment by scrubbbbs** [comment]
I think everyone is in agreement this needs to improve. I feel mouse controls need a refactor and should have their own settings section. We have various things that are bound to left/right/middle/back/forward/scroll mouse buttons, sometimes with modifier keys. We also do not have touch controls settings which are closely related.
